datasets/noaa-cdr/noaa_cdr.py

```diff
@@ -0,0 +1,189 @@
+
+
+class SeaIceConcentrationCollection(Collection):
+    """
+    As of 2023-03-31, noaa-cdr-sea-ice-concentration does not work. The data in
+    Azure blob storage is V2, but the data on the NOAA FTP site is V4, and the
+    stactools package was developed against V4. We're retaining this section for
+    the future when the Azure blob storage data is brought up-to-date.
+    """
+
+    @classmethod
+    def create_item(
+        cls, asset_uri: str, storage_factory: StorageFactory
+    ) -> Union[List[Item], WaitTaskResult]:
+        asset_storage, asset_path = storage_factory.get_storage_for_file(asset_uri)
+        year = asset_path.split("/")[1]
+        cog_storage = storage_factory.get_storage(cog_uri("sea-ice-concentration"))
+        with TemporaryDirectory() as temporary_directory:
+            file_name = os.path.basename(asset_path)
+            outfile = os.path.join(temporary_directory, file_name)
+            asset_storage.download_file(asset_path, outfile)
+            item = stactools.noaa_cdr.sea_ice_concentration.stac.create_item(outfile)
+            item.assets[
+                stactools.noaa_cdr.constants.NETCDF_ASSET_KEY
+            ].href = asset_storage.get_url(asset_path)
+
+            item = stactools.noaa_cdr.sea_ice_concentration.stac.add_cogs(
+                item, temporary_directory
+            )
+            for key, asset in item.assets.items():
+                if asset.media_type == MediaType.COG:
+                    cog_path = f"{year}/{os.path.basename(asset.href)}"
+                    cog_storage.upload_file(asset.href, cog_path)
+                    asset.href = cog_storage.get_url(cog_path)
+                    item.assets[key] = asset
+        return [item]
+
+
+class SeaSurfaceTemperatureOptimumInterpolationCollection(Collection):
+    @classmethod
+    def create_item(
+        cls, asset_uri: str, storage_factory: StorageFactory
+    ) -> Union[List[Item], WaitTaskResult]:
+        asset_storage, asset_path = storage_factory.get_storage_for_file(asset_uri)
+        if os.path.splitext(asset_path)[0].endswith("_preliminary"):
+            return []
+        prefix = asset_path.split("/")[-2]
+        cog_storage = storage_factory.get_storage(
+            cog_uri("sea-surface-temperature-optimum-interpolation")
+        )
+        with TemporaryDirectory() as temporary_directory:
+            file_name = os.path.basename(asset_path)
+            outfile = os.path.join(temporary_directory, file_name)
+            asset_storage.download_file(asset_path, outfile)
+            item = stactools.noaa_cdr.sea_surface_temperature_optimum_interpolation.stac.create_item(  # noqa
+                outfile
+            )
+            item.assets[
+                stactools.noaa_cdr.constants.NETCDF_ASSET_KEY
+            ].href = asset_storage.get_url(asset_path)
+
+            item = stactools.noaa_cdr.stac.add_cogs(item, temporary_directory)
+            for key, asset in item.assets.items():
```

Comment:
Noticing a bit of a pattern here (create item, create COGS, upload COGs), which signals the opportunity to consolidate some of this in a `pctasks` module. No necessary for this PR though.

Reply:
Yeah, the whole "are there cogs there already? no? create some, upload, and update the asset hrefs" pattern is relatively repeatable, but (as always) there's nuances (sometimes it's one netcdf -> many cogs, sometimes its many netcdfs -> many cogs).

I think some sort of "locally generated assets" mechanism could be an abstraction. "I made these things on my machine, pc-tasks can you please load them to this location and update my asset hrefs accordingly?"